### PR TITLE
Don't include type parameters in new preselection

### PR DESCRIPTION
**Pull request by rchande**
If a constructed generic type is inferred whose type arguments include type parameters, just provide the standard "completion of generic type" experience.
Fixes  https://github.com/dotnet/roslyn/issues/21049

### CODE REVIEW

**Comment by rchande** [comment]
Arguably, we could suggest `Task<T>` in the recursive case, but that doesn't really seem to add any value.

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.vb [hunk 1]
```diff
@@ -51,5 +51,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
         Protected Overrides Function GetCompletionItemRules(symbols As IReadOnlyList(Of ISymbol)) As CompletionItemRules
             Return s_rules
         End Function
+
+        Protected Overrides Function CreateItem(displayText As String, insertionText As String, symbols As List(Of ISymbol), context As SyntaxContext, preselect As Boolean, supportedPlatformData As SupportedPlatformData) As CompletionItem
```
Comment (dpoeschl) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.vb]: Unnecessary?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs [hunk 2]
```diff
@@ -79,6 +79,15 @@ protected override (string displayText, string insertionText) GetDisplayAndInser
                 return (symbol.Name, symbol.Name);
             }
 
+            // If not all  of the type arguments are concrete types, eg
+            // we've inferred something like `Task<T>`, just suggest the
+            // generic type `Task<>`
+            if (symbol is INamedTypeSymbol nt &&
+                nt.TypeArguments.Any(t => t.TypeKind == TypeKind.TypeParameter))
+            {
+                return CompletionUtilities.GetDisplayAndInsertionText(symbol, context);
```
Comment (dpoeschl) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: Am I supposed to know the difference between `CompletionUtilities.Stuff` and `base.Stuff` intuitively? I see that you're telling me that calling `CompletionUtilities.Stuff` will just do `Task<>` but I don't immediately understand why the difference. Probably just a matter of me not being familiar with this part of the code, but it's somewhat hard to read (I had to go digging around).

Comment (rchande) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: Maybe not intuitively, but probably some reading the code. The base method does some non-language specific stuff while the CompletionHelpers classes are language-specific. Perhaps something you would only know if you were familiar with completion.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs [hunk 3]
```diff
@@ -501,5 +501,46 @@ static void Main(string[] args)
...
+        Foo(new $$
+    }
+
+    void Bar()
+    {
+        Foo(new  Task<T>(null))
+    }
+}
+";
+            await VerifyItemExistsAsync(markup, "Task<>");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs]: why would i not want "Task<T>" after "Foo(new " here?  It's totally legal.

**Comment by CyrusNajmabadi** [comment]
21049 is a case where there is a reference to a *method's* type parameter from a different method.  But that's not hte case that's being checked in the code.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs [hunk 4]
```diff
@@ -79,9 +79,31 @@ protected override (string displayText, string insertionText) GetDisplayAndInser
...
+                nt.TypeArguments.Any(t => t is ITypeParameterSymbol typeParameter
+                    && !typeParameter.IsAccessibleFromSymbolOrEnclosingType(enclosingSymbol)))
+            {
+                return CompletionUtilities.GetDisplayAndInsertionText(symbol, context);
+            }
+
             return base.GetDisplayAndInsertionText(symbol, context);
         }
 
+        private bool IsApplicable(ITypeSymbol t, ISymbol enclosingSymbol)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: Is this called?

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeParameterSymbolExtensions.cs [hunk 5]
```diff
@@ -20,5 +20,15 @@ private static INamedTypeSymbol GetNamedTypeSymbol(ITypeSymbol type)
                     ? GetNamedTypeSymbolConstraint((ITypeParameterSymbol)type)
                     : null;
         }
+
+        public static bool IsAccessibleFromSymbolOrEnclosingType(this ITypeParameterSymbol typeParameter, ISymbol within)
+        {
+            if (typeParameter.DeclaringMethod != null)
+            {
+                return typeParameter.DeclaringMethod == within;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeParameterSymbolExtensions.cs]: you might be in a nested local function inside that method (add test for this plz).

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs [hunk 6]
```diff
@@ -79,6 +79,17 @@ protected override (string displayText, string insertionText) GetDisplayAndInser
                 return (symbol.Name, symbol.Name);
             }
 
+            // If not all  of the type arguments are concrete types, eg
+            // we've inferred something like `Task<T>`, just suggest the
+            // generic type `Task<>`
+            var enclosingSymbol = context.SemanticModel.GetEnclosingSymbol(context.Position);
+            if (symbol is INamedTypeSymbol nt &&
+                nt.TypeArguments.Any(t => t is ITypeParameterSymbol typeParameter
+                    && !typeParameter.IsAccessibleFromSymbolOrEnclosingType(enclosingSymbol)))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: :memo: The capture of `enclosingSymbol` means this method gets at least two new allocations.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: This feels *super* weird that this is the place we're updating any logic...

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs [hunk 7]
```diff
@@ -79,6 +79,17 @@ protected override (string displayText, string insertionText) GetDisplayAndInser
                 return (symbol.Name, symbol.Name);
             }
 
+            // If not all  of the type arguments are concrete types, eg
+            // we've inferred something like `Task<T>`, just suggest the
+            // generic type `Task<>`
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: this comment doesn't give te "why".  "why" would we want to do this?

It seems completely reasonable to both infer and suggest "Task<T>" if that's appropriate to the context we're in.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.cs]: this is the wrong place for this logic.  The place we should be doing this is where we're actually figuring out the symbol to preselect.  once we figure out htat symbol, we should update things *there* to say if we actually need to use another symbol.  We do not need change the "display/insertion text logic".

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.vb [hunk 8]
```diff
@@ -51,5 +51,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
...
         End Function
+
+        Protected Overrides Function GetDisplayAndInsertionText(symbol As ISymbol, context As SyntaxContext) As (displayText As String, insertionText As String)
+            ' If not all  of the type arguments are concrete types, eg
+            ' we've inferred something like `Task(Of T)`, just suggest the
+            ' generic `Task(Of ...)`
+            Dim enclosingSymbol = context.SemanticModel.GetEnclosingSymbol(context.Position)
+            Dim nt = TryCast(symbol, INamedTypeSymbol)
+            If nt.TypeArguments.Any(Function(t) t.TypeKind = TypeKind.TypeParameter AndAlso
+                                            Not DirectCast(t, ITypeParameterSymbol).IsAccessibleFromSymbolOrEnclosingType(enclosingSymbol)) Then
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/ObjectCreationCompletionProvider.vb]: 1. all my comments about c# apply here.
2. if your logic is identical, we should find a way to share this code between C# and VB in some helper method.

**Comment by CyrusNajmabadi** [comment]
It's not sufficient to just check the main type args.  Imagine if we inferred `List<T[]>` or `List<Task<T>>` where 'T' was a method-type-param for another method.  These checks would not work here.  Can you please add tests and update accordingly?

**Comment by CyrusNajmabadi** [comment]
Note: there are helpers to tell you which type parameters have been used by the entire type (recursively).  if you can't find them, i can help point you to them (i'm goin gon a plane soon, so i can't do it now).

**Comment by jinujoseph** [comment]
Closing this stale PR

**Pull request closed without merging**